// @deploysentinel/cypress-runner/dist/patch.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    patchServerInit: ()=>patchServerInit,
    patch: ()=>patch,
    run: ()=>run,
    getBinaryVersion: ()=>getBinaryVersion,
    verify: ()=>verify
});
const _childProcess = /*#__PURE__*/ _interopRequireDefault(require("child_process"));
const _fs = /*#__PURE__*/ _interopRequireDefault(require("fs"));
const _semver = /*#__PURE__*/ _interopRequireDefault(require("semver"));
const _jsYaml = /*#__PURE__*/ _interopRequireDefault(require("js-yaml"));
const _debug = require("./debug");
const _discovery = require("./discovery");
const _binPath = require("./bin-path");
const _js = require("./js");
const _fs1 = require("./fs");
function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) {
    try {
        var info = gen[key](arg);
        var value = info.value;
    } catch (error) {
        reject(error);
        return;
    }
    if (info.done) {
        resolve(value);
    } else {
        Promise.resolve(value).then(_next, _throw);
    }
}
function _asyncToGenerator(fn) {
    return function() {
        var self = this, args = arguments;
        return new Promise(function(resolve, reject) {
            var gen = fn.apply(self, args);
            function _next(value) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value);
            }
            function _throw(err) {
                asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err);
            }
            _next(undefined);
        });
    };
}
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function patchServerInit(injectedAbsolutePath) {
    return _patchServerInit.apply(this, arguments);
}
function _patchServerInit() {
    _patchServerInit = _asyncToGenerator(function*(injectedAbsolutePath) {
        const serverInitPath = yield (0, _discovery.getServerInitPath)();
        (0, _debug.debug)('Patching cypress entry point file: %s', serverInitPath);
        const serverInitBackup = serverInitPath + '.bak';
        if (!(0, _fs1.pathExists)(serverInitBackup)) {
            _fs.default.copyFileSync(serverInitPath, serverInitBackup);
        }
        const doc = _fs.default.readFileSync(serverInitPath, 'utf8');
        const result = (0, _js.instrumentCypressInit)(doc, injectedAbsolutePath, serverInitPath, serverInitBackup);
        _fs.default.writeFileSync(serverInitPath, result);
    });
    return _patchServerInit.apply(this, arguments);
}
function patch(apiURL, cypressConfigFilePath) {
    return _patch.apply(this, arguments);
}
function _patch() {
    _patch = _asyncToGenerator(function*(apiURL, cypressConfigFilePath) {
        if (!apiURL) {
            throw new Error('Missing apiURL');
        }
        const { configFilePath , backupConfigFilePath  } = yield (0, _discovery.getConfigFilesPaths)(cypressConfigFilePath);
        (0, _debug.debug)('Patching cypress config file: %s', configFilePath);
        try {
            _fs.default.statSync(backupConfigFilePath);
        } catch (e) {
            _fs.default.copyFileSync(configFilePath, backupConfigFilePath);
        }
        const doc = _jsYaml.default.load(_fs.default.readFileSync(configFilePath, 'utf8'));
        doc.production.api_url = apiURL;
        _fs.default.writeFileSync(configFilePath, _jsYaml.default.dump(doc));
    });
    return _patch.apply(this, arguments);
}
function run() {
    return _run.apply(this, arguments);
}
function _run() {
    _run = _asyncToGenerator(function*() {
        const [node, script, ...rest] = process.argv;
        const cliBinPath = yield (0, _binPath.getCypressCLIBinPath)();
        (0, _debug.debug)('Running cypress from %s', cliBinPath, ...rest);
        return _childProcess.default.spawn(cliBinPath, [
            ...rest
        ], {
            stdio: 'inherit'
        });
    });
    return _run.apply(this, arguments);
}
function getBinaryVersion() {
    return _getBinaryVersion.apply(this, arguments);
}
function _getBinaryVersion() {
    _getBinaryVersion = _asyncToGenerator(function*() {
        const cliBinPath = yield (0, _binPath.getCypressCLIBinPath)();
        (0, _debug.debug)('Checking cypress version from %s', cliBinPath);
        try {
            const stdout = _childProcess.default.execFileSync(cliBinPath, [
                'version'
            ], {
                stdio: 'pipe'
            }).toString();
            const lines = stdout.split('\n');
            for (const line of lines){
                if (line.includes('Cypress binary version: ')) {
                    const version = line.replace('Cypress binary version: ', '').trim();
                    if (_semver.default.valid(version)) {
                        return version;
                    }
                }
            }
        } catch (e) {
            (0, _debug.debug)(e);
            return null;
        }
    });
    return _getBinaryVersion.apply(this, arguments);
}
function verify() {
    return _verify.apply(this, arguments);
}
function _verify() {
    _verify = _asyncToGenerator(function*() {
        const cliBinPath = yield (0, _binPath.getCypressCLIBinPath)();
        (0, _debug.debug)('Verifying cypress from %s', cliBinPath);
        _childProcess.default.execFileSync(cliBinPath, [
            'verify'
        ], {
            stdio: 'pipe'
        });
    });
    return _verify.apply(this, arguments);
}
